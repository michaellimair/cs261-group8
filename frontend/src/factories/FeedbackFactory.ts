/* eslint-disable class-methods-use-this */
import faker from '@faker-js/faker';
import { FeedbackType, IFeedback } from 'customTypes/feedback';
import { merge } from 'lodash';
import FeedbackReplyFactory from './FeedbackReplyFactory';

class FeedbackFactory {
  create = (feedback?: Partial<IFeedback>): IFeedback => {
    const replyFactory = new FeedbackReplyFactory();
    return merge({
      id: faker.datatype.number(),
      title: faker.lorem.lines(1),
      content: faker.lorem.paragraph(3),
      type: FeedbackType.BUG,
      created: new Date(),
      modified: new Date(),
      reply: replyFactory.create(),
    }, feedback);
  };

  createMany = (count: number) => Array(count).fill(null).map((_, i) => this.create({ id: i }));
}

export default FeedbackFactory;
