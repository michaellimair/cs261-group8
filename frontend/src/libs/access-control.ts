import { IUserGroup, UserGroup } from 'customTypes/auth';

/** Function to check whether a user is allowed to access a certain resource
 * given the list of groups which can access the resource. */
export const checkAllowed = ({
  userGroups,
  allowedGroups,
}: {
  userGroups: IUserGroup[];
  allowedGroups: UserGroup[];
}): boolean => {
  if (allowedGroups.includes(UserGroup.UNDECLARED) && !userGroups.length) {
    return true;
  }
  return allowedGroups
    .some((groupName) => userGroups.find(({ name }) => name === groupName));
};
