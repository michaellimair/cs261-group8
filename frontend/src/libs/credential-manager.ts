export const JWT_TOKEN_KEY = 'jwt_token';
export const JWT_EXPIRY_KEY = 'jwt_expiry';

class CredentialManager {
  protected token: string | null = null;

  protected expiry: Date | null = null;

  constructor(private readonly storage: Storage = localStorage) {
    this.restoreAuthenticationCredentials();
  }

  private restoreAuthenticationCredentials() {
    this.token = this.storage.getItem(JWT_TOKEN_KEY);
    const expiryStr = this.storage.getItem(JWT_EXPIRY_KEY);

    if (expiryStr) {
      this.expiry = new Date(expiryStr);
    }

    if (this.isCredentialExpired()) {
      this.clearCredentials();
    }
  }

  private isCredentialExpired(): boolean {
    return this.expiry !== null && (new Date()).getTime() > this.expiry.getTime();
  }

  setCredentials(token: string, expiry: Date) {
    this.token = token;
    this.expiry = expiry;
    this.storage.setItem(JWT_TOKEN_KEY, this.token);
    this.storage.setItem(JWT_EXPIRY_KEY, this.expiry.toISOString());
  }

  clearCredentials() {
    this.storage.removeItem(JWT_TOKEN_KEY);
    this.storage.removeItem(JWT_EXPIRY_KEY);

    this.token = null;
    this.expiry = null;
  }

  get credentials() {
    if (this.isCredentialExpired()) {
      this.clearCredentials();
    }
    return {
      token: this.token,
      expiry: this.expiry,
    };
  }
}

export default CredentialManager;
