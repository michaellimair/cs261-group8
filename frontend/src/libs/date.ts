export const formatDateTime = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', {
  weekday: 'long',
  timeZoneName: 'short',
  day: 'numeric',
  month: 'long',
  year: 'numeric',
  hour: 'numeric',
  minute: 'numeric',
  timeZone,
}).format(date);

export const formatTime = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', { hour: 'numeric', minute: 'numeric', timeZone }).format(date);

export const formatDate = (date: Date, timeZone?: string): string => new Intl.DateTimeFormat('en-GB', {
  day: 'numeric', month: 'short', year: 'numeric', timeZone,
}).format(date);
