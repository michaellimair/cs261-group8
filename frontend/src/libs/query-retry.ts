import NotFoundError from 'api/error/NotFoundError';
import UnauthorizedError from 'api/error/UnauthorizedError';
import { RetryValue } from 'react-query/types/core/retryer';

export const queryRetryCondition: Exclude<RetryValue<unknown>, number | boolean> = (
  retryCount,
  err,
) => retryCount < 3
    && !NotFoundError.isNotFoundError(err)
    && !UnauthorizedError.isUnauthorizedError(err);
