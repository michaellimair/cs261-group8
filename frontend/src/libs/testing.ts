export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});
