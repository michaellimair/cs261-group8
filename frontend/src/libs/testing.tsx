import { FC } from 'react';

import {
  MemoryRouter, Route, Routes,
} from 'react-router-dom';

export const tick = (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

export const QueryRouterWrapper: FC = ({ children }) => (
  <MemoryRouter initialEntries={['/']}>
    <Routes>
      <Route path="*" element={children} />
    </Routes>
  </MemoryRouter>
);
