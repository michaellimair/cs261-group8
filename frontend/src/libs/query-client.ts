import { QueryClient } from 'react-query';
import { queryRetryCondition } from 'libs/query-retry';

export const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      // If error is not a 404 error, allow retry
      retry: queryRetryCondition,
      cacheTime: 0,
    },
  },
});
