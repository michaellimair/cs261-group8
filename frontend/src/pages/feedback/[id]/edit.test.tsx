import { render } from '@testing-library/react';
import FeedbackFactory from 'factories/FeedbackFactory';
import { QueryRouterWrapper } from 'libs/testing';
import { useQuery } from 'react-query';
import EditFeedbackPage from './edit';

jest.mock('react-query', () => ({
  useQuery: jest.fn(),
  useMutation: jest.fn(),
}));

const mockReset = jest.fn();

jest.mock('hooks/useCommonForm', () => () => ({
  onSubmit: jest.fn(),
  errors: {},
  isLoading: false,
  register: jest.fn(),
  reset: mockReset,
}));

describe('pages/feedback/[id]/edit', () => {
  it('shows edit form properly', () => {
    const feedbackFactory = new FeedbackFactory();
    const data = feedbackFactory.create();
    (useQuery as jest.Mock).mockImplementation(() => ({
      isLoading: false,
      data,
    }));

    const result = render(
      <QueryRouterWrapper>
        <EditFeedbackPage />
      </QueryRouterWrapper>,
    );

    expect(mockReset).toHaveBeenCalledTimes(1);
    expect(result).toMatchSnapshot();
  });
});
