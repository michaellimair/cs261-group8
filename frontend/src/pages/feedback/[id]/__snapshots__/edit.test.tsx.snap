// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/feedback/[id]/edit shows edit form properly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
    </div>
    <div>
      <div
        class="chakra-stack css-n21gh5"
      >
        <h2
          class="chakra-heading css-0"
        >
          dashboard.edit_feedback.title
        </h2>
        <div
          class="css-jhpxgx"
        >
          <form
            data-testid="feedbackForm"
          >
            <div
              class="chakra-form-control css-wlea3r"
              data-testid="type-formcontrol"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                data-testid="type-label"
                for="type"
                id="type-label"
              >
                dashboard.feedback.type
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <div
                class="chakra-select__wrapper css-42b2qy"
              >
                <select
                  aria-required="true"
                  class="chakra-select css-1ik61og"
                  id="type"
                  required=""
                >
                  <option
                    value="bug"
                  >
                    dashboard.feedback.bug
                  </option>
                  <option
                    value="improvement"
                  >
                    dashboard.feedback.improvement
                  </option>
                  <option
                    value="question"
                  >
                    dashboard.feedback.question
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-form-control css-wlea3r"
              data-testid="title-formcontrol"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                data-testid="title-label"
                for="title"
                id="title-label"
              >
                title
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <div
                class="chakra-input__group css-4302v8"
              >
                <input
                  aria-required="true"
                  class="chakra-input css-0"
                  data-testid="title"
                  id="title"
                  required=""
                  type="text"
                />
              </div>
            </div>
            <div
              class="chakra-form-control css-wlea3r"
              data-testid="content-formcontrol"
              role="group"
            >
              <label
                class="chakra-form__label css-uvho5g"
                data-testid="content-label"
                for="content"
                id="content-label"
              >
                content
                <span
                  aria-hidden="true"
                  class="chakra-form__required-indicator css-0"
                  role="presentation"
                >
                  *
                </span>
              </label>
              <textarea
                aria-required="true"
                class="chakra-textarea css-0"
                id="content"
                required=""
              />
            </div>
            <div
              class="chakra-form-control css-t9xfd"
              role="group"
            />
            <div
              class="chakra-stack css-17vvmxc"
            >
              <button
                class="chakra-button css-5c2176"
                data-testid="edit-feedback-button"
                type="submit"
              >
                dashboard.edit_feedback.title
              </button>
              <a
                class="chakra-link chakra-button css-ogxy0s"
                href="/feedbacks/undefined"
              >
                dashboard.edit_feedback.cancel
              </a>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="chakra-stack css-n21gh5"
    >
      <h2
        class="chakra-heading css-0"
      >
        dashboard.edit_feedback.title
      </h2>
      <div
        class="css-jhpxgx"
      >
        <form
          data-testid="feedbackForm"
        >
          <div
            class="chakra-form-control css-wlea3r"
            data-testid="type-formcontrol"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              data-testid="type-label"
              for="type"
              id="type-label"
            >
              dashboard.feedback.type
              <span
                aria-hidden="true"
                class="chakra-form__required-indicator css-0"
                role="presentation"
              >
                *
              </span>
            </label>
            <div
              class="chakra-select__wrapper css-42b2qy"
            >
              <select
                aria-required="true"
                class="chakra-select css-1ik61og"
                id="type"
                required=""
              >
                <option
                  value="bug"
                >
                  dashboard.feedback.bug
                </option>
                <option
                  value="improvement"
                >
                  dashboard.feedback.improvement
                </option>
                <option
                  value="question"
                >
                  dashboard.feedback.question
                </option>
              </select>
              <div
                class="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-select__icon"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="chakra-form-control css-wlea3r"
            data-testid="title-formcontrol"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              data-testid="title-label"
              for="title"
              id="title-label"
            >
              title
              <span
                aria-hidden="true"
                class="chakra-form__required-indicator css-0"
                role="presentation"
              >
                *
              </span>
            </label>
            <div
              class="chakra-input__group css-4302v8"
            >
              <input
                aria-required="true"
                class="chakra-input css-0"
                data-testid="title"
                id="title"
                required=""
                type="text"
              />
            </div>
          </div>
          <div
            class="chakra-form-control css-wlea3r"
            data-testid="content-formcontrol"
            role="group"
          >
            <label
              class="chakra-form__label css-uvho5g"
              data-testid="content-label"
              for="content"
              id="content-label"
            >
              content
              <span
                aria-hidden="true"
                class="chakra-form__required-indicator css-0"
                role="presentation"
              >
                *
              </span>
            </label>
            <textarea
              aria-required="true"
              class="chakra-textarea css-0"
              id="content"
              required=""
            />
          </div>
          <div
            class="chakra-form-control css-t9xfd"
            role="group"
          />
          <div
            class="chakra-stack css-17vvmxc"
          >
            <button
              class="chakra-button css-5c2176"
              data-testid="edit-feedback-button"
              type="submit"
            >
              dashboard.edit_feedback.title
            </button>
            <a
              class="chakra-link chakra-button css-ogxy0s"
              href="/feedbacks/undefined"
            >
              dashboard.edit_feedback.cancel
            </a>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
